vana/utils/transaction.py

```diff
@@ -0,0 +1,167 @@
+from typing import Optional, Tuple, Dict, Any
+from web3 import Web3
+from web3.types import TxReceipt, HexBytes
+from eth_account.signers.local import LocalAccount
+import time
+import vana
+
+class TransactionManager:
+    def __init__(self, web3: Web3, account: LocalAccount):
+        self.web3 = web3
+        self.account = account
+        self._nonce_cache: Dict[str, int] = {}
+        self._last_nonce_refresh = 0
+        self.nonce_refresh_interval = 60
+        self.chain_id = self.web3.eth.chain_id
+
+    def _get_safe_nonce(self) -> int:
+        """
+        Get the next safe nonce, accounting for pending transactions
+        """
+        current_time = time.time()
+        cache_key = self.account.address
+
+        # Refresh nonce cache if expired
+        if current_time - self._last_nonce_refresh > self.nonce_refresh_interval:
+            # Get the latest confirmed nonce
+            confirmed_nonce = self.web3.eth.get_transaction_count(self.account.address, 'latest')
+            # Get pending nonce
+            pending_nonce = self.web3.eth.get_transaction_count(self.account.address, 'pending')
+            # Use the higher of the two to avoid nonce conflicts
+            self._nonce_cache[cache_key] = max(confirmed_nonce, pending_nonce)
+            self._last_nonce_refresh = current_time
+
+        # Get and increment the cached nonce
+        nonce = self._nonce_cache.get(cache_key, 0)
+        self._nonce_cache[cache_key] = nonce + 1
+        return nonce
+
+    def _clear_pending_transactions(self):
+        """
+        Clear pending transactions by sending zero-value transactions with higher gas price
+        """
+        try:
+            # Get all pending transactions for the account
+            pending_nonce = self.web3.eth.get_transaction_count(self.account.address, 'pending')
+            confirmed_nonce = self.web3.eth.get_transaction_count(self.account.address, 'latest')
+
+            if pending_nonce > confirmed_nonce:
+                vana.logging.info(f"Clearing {pending_nonce - confirmed_nonce} pending transactions")
+
+                # Send replacement transactions with higher gas price
+                for nonce in range(confirmed_nonce, pending_nonce):
+                    replacement_tx = {
+                        'from': self.account.address,
+                        'to': self.account.address,
+                        'value': 0,
+                        'nonce': nonce,
+                        'gas': 21000,
```

Comment:
Where does `21000` come from?

Reply:
21000 is the standard gas cost for a basic ETH transfer transaction in Ethereum - this is a fixed amount that represents the minimum computational work needed to process a simple transfer of ETH from one address to another.

This is specifically used in _clear_pending_transactions() because we're sending simple ETH transfers (value 0) to the same address - these are the simplest possible transactions that require exactly 21000 gas.

Reply:
Made this clearer in the code with a constant